src/providers/aliyun/mod.rs

```diff
@@ -0,0 +1,187 @@
+//! Alibaba Cloud
+//!
+//! This provider is selected via the platform ID `aliyun`.
+//! The metadata endpoint is documented at https://www.alibabacloud.com/help/doc-detail/49122.htm.
+
+#[cfg(test)]
+use mockito;
+use openssh_keys::PublicKey;
+use std::collections::{BTreeSet, HashMap};
+
+use crate::errors::*;
+use crate::network;
+use crate::providers::MetadataProvider;
+use crate::retry;
+
+#[cfg(test)]
+mod mock_tests;
+
+/// Provider prefix for Alibaba Cloud.
+static PROVIDER_PREFIX: &str = "ALIYUN";
+
+#[derive(Clone, Debug)]
+pub struct AliyunProvider {
+    client: retry::Client,
+}
+
+impl AliyunProvider {
+    pub fn try_new() -> Result<AliyunProvider> {
+        let client = retry::Client::try_new()?.return_on_404(true);
+
+        Ok(AliyunProvider { client })
+    }
+
+    #[cfg(test)]
+    fn endpoint_for(name: &str) -> String {
+        let url = mockito::server_url();
+        format!("{}/{}", url, name)
+    }
+
+    #[cfg(not(test))]
+    fn endpoint_for(name: &str) -> String {
+        format!("http://100.100.100.200/latest/meta-data/{}", name)
+    }
+
+    /// Fetch a metadata attribute from its specific endpoint.
+    ///
+    /// Content (if any) is stored into the provided `map`,
+    /// overwriting any previous existing value under the same `key`.
+    fn fetch_attribute(
+        &self,
+        map: &mut HashMap<String, String>,
+        key: &str,
+        endpoint: &str,
+    ) -> Result<()> {
+        let content: Option<String> = self
+            .client
+            .get(retry::Raw, Self::endpoint_for(endpoint))
+            .send()?;
+
+        if let Some(value) = content {
+            if !value.is_empty() {
+                map.insert(key.to_string(), value);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Retrieve SSH public keys.
+    ///
+    /// Note: this uses a `BTreeSet` to de-duplicate redundant
+    /// entries returned by the metadata API.
+    fn fetch_ssh_keys(&self) -> Result<BTreeSet<String>> {
+        let entries: Option<String> = self
+            .client
+            .get(retry::Raw, AliyunProvider::endpoint_for("public-keys/"))
+            .send()?;
+        let keys_list = entries.unwrap_or_default();
+
+        let mut out = BTreeSet::new();
+        for entry in keys_list.lines() {
+            let key_id = entry.trim_end_matches('/');
+            let ep = format!("public-keys/{}/openssh-key", key_id);
```

Comment:
Just curious: Did you find the single key endpoint via trial and error or was it documented? A quick look at the docs and using search didn't turn it up.

Reply:
It is not documented so I had to do a bit of experiments.
However it is basically a 1:1 of the AWS logic: https://github.com/coreos/afterburn/blob/59d3c56ccefe093ec98d5aecb2579b88b7b5e9b5/src/providers/aws/mod.rs#L88.
For that reason, I only documented the "duplicated entry" quirk.

Reply:
It's also "documented" in the cloud-init implementation :)
https://github.com/cloud-init/cloud-init/commit/4f8ceffb2e3a9feefcb718bda7a7f0f21ef7ab7c